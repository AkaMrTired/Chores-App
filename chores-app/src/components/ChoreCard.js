/* eslint-disable object-shorthand */
/* eslint-disable react/prop-types */
// not all props required all the time
import React, { useState } from "react";
import axios from "axios";
import PropTypes from "prop-types";
import EditChore from "./EditChore";

import { useUserAuth } from "../context/UserAuthContext";

const ChoreCard = ({ name, price, status, component, choreID, choreOwner }) => {
  const [editing, setEditing] = useState(false);
  const { setChores } = useUserAuth();
  const familyID = localStorage.getItem("familyID");
  const userID = localStorage.getItem("userID");

  const editButton = (event) => {
    event.preventDefault();
    setEditing(true);
  };

  const deleteButton = (event) => {
    event.preventDefault();
    axios
      .delete(`http://localhost:3300/family/${familyID}/chores/${choreID}`)
      .then(() => {
        return axios.get(`http://localhost:3300/family/${familyID}/chores`);
      })
      .then((response) => {
        setChores(response.data);
      })
      .catch((e) => {
        console.log(e);
      });
  };

  const acceptButton = (event) => {
    event.preventDefault();
    axios
      .patch(`http://localhost:3300/family/${familyID}/chores/${choreID}`, {
        status: "U",
        owner: null,
      })
      .then(() => {
        return axios.get(`http://localhost:3300/family/${familyID}/chores`);
      })
      .then((response) => {
        setChores(response.data);
      })
      .catch((e) => {
        console.log(e);
      });
    // FUTURE DEVELOPMENT:
    // -> money goes into child's balance
  };
  const rejectButton = (event) => {
    event.preventDefault();
    axios
      .patch(`http://localhost:3300/family/${familyID}/chores/${choreID}`, {
        status: "T",
        owner: choreOwner,
      })
      .then(() => {
        return axios.get(`http://localhost:3300/family/${familyID}/chores`);
      })
      .then((response) => {
        setChores(response.data);
      })
      .catch((e) => {
        console.log(e);
      });
  };
  const doneButton = (event) => {
    event.preventDefault();
    axios
      .patch(`http://localhost:3300/family/${familyID}/chores/${choreID}`, {
        status: "P",
        owner: choreOwner,
      })
      .then(() => {
        return axios.get(`http://localhost:3300/family/${familyID}/chores`);
      })
      .then((response) => {
        setChores(response.data);
      })
      .catch((e) => {
        console.log(e);
      });
  };
  const takeButton = (event) => {
    event.preventDefault();
    axios
      .patch(`http://localhost:3300/family/${familyID}/chores/${choreID}`, {
        status: "T",
        owner: userID,
      })
      .then(() => {
        return axios.get(`http://localhost:3300/family/${familyID}/chores`);
      })
      .then((response) => {
        setChores(response.data);
      })
      .catch((e) => {
        console.log(e);
      });
  };

  if (editing) {
    return (
      <EditChore
        name={name}
        price={price}
        status={status}
        setEditing={setEditing}
        choreID={choreID}
        owner={choreOwner}
      />
    );
  }
  return (
    <div>
      <div>{name}</div>
      <div>Â£{price}</div>
      {!!status && <div>{status}</div>}
      {component === "ParentDashboard" && (
        <button type="button" onClick={editButton}>
          Edit
        </button>
      )}
      {component === "ParentDashboard" && (
        <button type="button" onClick={deleteButton}>
          Delete
        </button>
      )}
      {component === "ChoresToApprove" && (
        <button type="button" onClick={acceptButton}>
          Accept
        </button>
      )}
      {component === "ChoresToApprove" && (
        <button type="button" onClick={rejectButton}>
          Reject
        </button>
      )}
      {component === "ChildDashboard" && (
        <button type="button" onClick={doneButton}>
          Done
        </button>
      )}
      {component === "FindAvailableChores" && (
        <button type="button" onClick={takeButton}>
          Take
        </button>
      )}
    </div>
  );
};

ChoreCard.propTypes = {
  name: PropTypes.string.isRequired,
  price: PropTypes.number.isRequired,
};

export default ChoreCard;
