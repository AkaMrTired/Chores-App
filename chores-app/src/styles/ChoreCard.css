.chore-card {
  display: flex;
  flex-direction: column;
  padding: 1rem;
  margin: 0 0.5rem 2rem;
  background-color: var(--white);
  border-radius: 15px;
  box-shadow: var(--shadow);
  align-self: center;
  width: 100%;
  max-width: 500px;
}

.chore-btn {
  border-radius: 15px;
  font-family: inherit;
  padding: 0.5rem 1rem;
}

.chore-btn:first-of-type {
  margin: 0 0.5rem 0 1rem;
}

.chore-btn_pink {
  background-color: var(--pink);
  border: 1px solid var(--pink);
}

.chore-btn_pink:hover {
  background-color: var(--pink-dark);
  border: 1px solid var(--pink-dark);
}

.chore-btn_peach {
  background-color: var(--peach);
  border: 1px solid var(--peach);
}

.chore-btn_peach:hover {
  background-color: var(--peach-dark);
  border: 1px solid var(--peach-dark);
}

.chore-price {
  background-color: var(--purple-dark);
  color: var(--white);
  padding: 0.5rem 1rem;
  border-radius: 25px;
  line-height: 1.7;
  margin-right: 1rem;
}

.chore-name {
  width: 100%;
  text-align: center;
  padding: 0.5rem 0;
  background-color: var(--purple);
  border-radius: 15px;
  margin-bottom: 1rem;
}

.chore-buttons-wrapper {
  display: flex;
  justify-content: space-around;
  align-items: center;
}
