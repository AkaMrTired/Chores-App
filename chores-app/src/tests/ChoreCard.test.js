import React from "react";
import { render, screen } from "@testing-library/react";
import ChoreCard from "../components/ChoreCard";

const validProps = {
  name: "test",
  price: "1",
  status: "T",
  emptyFunction: () => {},
};

describe("the chore card component", () => {
  it("matches the screenshot", () => {
    const { asFragment } = render(
      <ChoreCard
        name={validProps.name}
        price={validProps.price}
        status={validProps.status}
        editButton={validProps.emptyFunction}
        deleteButton={validProps.emptyFunction}
        acceptButton={validProps.emptyFunction}
        rejectButton={validProps.emptyFunction}
        doneButton={validProps.emptyFunction}
        takeButton={validProps.emptyFunction}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders all the components when provided with valid validProps", () => {
    render(
      <ChoreCard
        name={validProps.name}
        price={validProps.price}
        status={validProps.status}
        editButton={validProps.emptyFunction}
        deleteButton={validProps.emptyFunction}
        acceptButton={validProps.emptyFunction}
        rejectButton={validProps.emptyFunction}
        doneButton={validProps.emptyFunction}
        takeButton={validProps.emptyFunction}
      />
    );
    expect(screen.getByText("test")).toBeInTheDocument();
    expect(screen.getByText("£1")).toBeInTheDocument();
    expect(screen.getByText("T")).toBeInTheDocument();
    expect(screen.getByText("edit")).toBeInTheDocument();
    expect(screen.getByText("delete")).toBeInTheDocument();
    expect(screen.getByText("accept")).toBeInTheDocument();
    expect(screen.getByText("reject")).toBeInTheDocument();
    expect(screen.getByText("done")).toBeInTheDocument();
    expect(screen.getByText("take")).toBeInTheDocument();
  });
  it("doesn't render components when it isn't given the relevant prop", () => {
    render(<ChoreCard name={validProps.name} price={validProps.price} />);
    expect(screen.getByText("test")).toBeInTheDocument();
    expect(screen.getByText("£1")).toBeInTheDocument();
    expect(screen.queryByText("T")).not.toBeInTheDocument();
    expect(screen.queryByText("edit")).not.toBeInTheDocument();
    expect(screen.queryByText("delete")).not.toBeInTheDocument();
    expect(screen.queryByText("accept")).not.toBeInTheDocument();
    expect(screen.queryByText("reject")).not.toBeInTheDocument();
    expect(screen.queryByText("done")).not.toBeInTheDocument();
    expect(screen.queryByText("take")).not.toBeInTheDocument();
  });
});
