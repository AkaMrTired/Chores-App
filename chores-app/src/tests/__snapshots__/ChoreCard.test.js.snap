// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the chore card component matches the screenshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      test
    </div>
    <div>
      Â£1
    </div>
    <div>
      T
    </div>
    <button
      type="button"
    >
      edit
    </button>
    <button
      type="button"
    >
      delete
    </button>
    <button
      type="button"
    >
      accept
    </button>
    <button
      type="button"
    >
      reject
    </button>
    <button
      type="button"
    >
      done
    </button>
    <button
      type="button"
    >
      take
    </button>
  </div>
</DocumentFragment>
`;
